RFS/src/main/java/org/opensearch/migrations/bulkload/delta/DeltaDocumentReaderEngine.java

```diff
@@ -53,24 +58,65 @@ public SnapshotShardUnpacker createUnpacker(
         );
     }
 
+    @SneakyThrows
+    private static void closeReaders(LuceneDirectoryReader first, LuceneDirectoryReader second) {
+        try {
+            if (first != null) {
+                first.close();
+            }
+        }
+        finally {
+            if (second != null) {
+                second.close();
+            }
+        }
+    }
+
     @Override
     public Flux<RfsLuceneDocument> readDocuments(
         LuceneIndexReader reader,
         String indexName,
         int shardNumber,
         int startingDocId,
         BaseRootRfsContext rootContext
-    ) {
+    ) throws IOException {
         ShardMetadata previousShardMetadata = previousShardMetadataFactory.apply(indexName, shardNumber);
         ShardMetadata shardMetadata = shardMetadataFactory.apply(indexName, shardNumber);
-        if (deltaMode != DeltaMode.UPDATES_ONLY) {
-            throw new UnsupportedOperationException("Unsupported delta mode given " + deltaMode);
+        LuceneDirectoryReader previousReader = null;
+        LuceneDirectoryReader currentReader = null;
+        try {
+            previousReader = reader.getReader(previousShardMetadata.getSegmentFileName());
+            currentReader  = reader.getReader(shardMetadata.getSegmentFileName());
+
+            var deltaResult = DeltaLuceneReader.readDeltaDocsByLeavesFromStartingPosition(
+                previousReader, currentReader, startingDocId, rootContext);
+
+            LuceneDirectoryReader finalPreviousReader = previousReader;
+            LuceneDirectoryReader finalCurrentReader = currentReader;
+            Runnable cleanup = () -> closeReaders(finalPreviousReader, finalCurrentReader);
+
+            if (DeltaMode.UPDATES_ONLY.equals(deltaMode)) {
+                return deltaResult.additions
+                    .doFinally(s -> cleanup.run());
+            } else if (DeltaMode.UPDATES_AND_DELETES.equals(deltaMode)) {
+                // Using concat to perform deletions first then additions
+                return Flux.concat(
+                    deltaResult.deletions,
+                    deltaResult.additions
+                ).doFinally(s -> cleanup.run());
+            } else if (DeltaMode.DELETES_ONLY.equals(deltaMode)) {
+                return deltaResult.deletions
+                    .doFinally(s -> cleanup.run());
+            } else {
+                throw new UnsupportedOperationException("Unsupported delta mode given " + deltaMode);
```

Comment:
Since this is an a try block it will be caught on 113 which is weird.  I think this should be removable by using a switch/case...  it will throw a complier error if an new enum types are added but not supported.

Reply:
In general input parameter should be inspected before the 'meat' of a method to make it easier to know inputs/outputs.